### PR TITLE
Adds binary encoding of TypeSection

### PR DESCRIPTION
**Pull request by codefromthecrypt**

### CODE REVIEW

File: wasm/type.go [hunk 1]
```diff
@@ -11,6 +11,30 @@ type FunctionType struct {
...
+var nullary = []byte{0x60, 0, 0}
+
+var encodedOneParam = buildEncodedFuncTypes(true)
+var encodedOneResult = buildEncodedFuncTypes(false)
+
+// buildEncodedFuncTypes is like buildEncodedValTypes except it is parameter or result specific and includes the func
+// type prefix 0x60.
+//
+// See https://www.w3.org/TR/wasm-core-1/#function-types%E2%91%A4
+func buildEncodedFuncTypes(param bool) (encodedTypes [ValueTypeI32 + 1][]byte) {
```
Comment (mathetake) [inline wasm/type.go:23]: `[ValueTypeI32 + 1][]byte` seems too big and redundant because `ValueTypeI32 = 0x7f`?

Comment (mathetake) [reply wasm/type.go:23]: maybe `ValueTypeF64 - ValueTypeI32`

File: wasm/value.go [hunk 2]
```diff
@@ -35,6 +35,42 @@ func formatValueType(t ValueType) (ret string) {
...
 
+var noValType = []byte{0}
+
+// encodedValTypes is a cache of size prefixed binary encoding of known val types.
+var encodedValTypes = buildEncodedValTypes()
+
+// buildEncodedValTypes builds results for encodeValTypes for known value types.
+//
+// Note: this is length ValueTypeI32+1 because the largest known val type is that.
+func buildEncodedValTypes() (encodedTypes [ValueTypeI32 + 1][]byte) {
```
Comment (mathetake) [inline wasm/value.go:46]: same

File: wasm/type.go (hunk 1 above)
Comment (codefromthecrypt) [reply wasm/type.go:23]: yeah we can make it smaller, but then to access the map entry, we have subtract first

File: wasm/value.go (hunk 2 above)
Comment (codefromthecrypt) [reply wasm/value.go:46]: it will be a bit slower, but sure!

File: wasm/type.go (hunk 1 above)
Comment (codefromthecrypt) [reply wasm/type.go:23]: at this point probably just as well to use a switch statement.

File: wasm/value.go (hunk 2 above)
Comment (codefromthecrypt) [reply wasm/value.go]: done!

File: wasm/value.go [hunk 3]
```diff
@@ -19,6 +19,13 @@ const (
 	ValueTypeI64 ValueType = 0x7e
 	ValueTypeF32 ValueType = 0x7d
 	ValueTypeF64 ValueType = 0x7c
+	// valTypeRange allows us to cache common valueType calculations. To determine the array offset, subtract
+	// valTypeFloor from the value of interest.
+	valTypeRange = ValueTypeI32 - valTypeFloor + 1
+	// valTypeFloor is the lowest numeric value of a value type.
+	//
+	// Note: the floor value changes after WebAssembly 1.0 (MVP)
```
Comment (codefromthecrypt) [inline wasm/value.go:27]: ps only reason I added this is I looked it up before punting to a larger array before, as it changes to https://webassembly.github.io/spec/core/binary/types.html#reference-types

I don't care about it, just made a comment to remind later.

**Review by mathetake — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
I redid this as a map, because it is the more natural type and by the time we start checking ranges, etc, the other approach started falling apart.

**Review by mathetake — APPROVED** [review]
I like it better!

**Comment by codefromthecrypt** [comment]
Thanks for your patience and advice @mathetake!

**Pull request merged**